File: CPP/Palindrome Reorder.cpp
#include <bits/stdc++.h>
using namespace std;

int main()
{
    string s;
    cin >> s;
    int n = s.size();
    
    unordered_map<char,int> m;
    int odd = 0;
    for(auto i : s){
        if(m.find(i) == m.end()){
            m[i] = count(s.begin(),s.end(),i);
            if(m[i]%2!=0 ){
                odd++;
                
            }
        }
    }
    if(odd > 1){
        cout << "NO SOLUTION";
    }
    else{
    int i = 0;
    for(auto it = m.begin();it != m.end();it++){
        if(it->second!=1){
            while(it->second!=0){
                if(it->second == 1){
                    s[(n-1)/2] = it->first;
                    it->second--;
                }
                else{
                s[i] = it->first;
                s[n - i - 1] = it->first;
                it->second = it->second - 2;
                i++;
                }
            }
        }
        else{
            s[(n-1)/2] = it->first;
        }

    }
    cout << s;
    }
    
    
    return 0;
}

File: PYTHON/Palindrome Reorder.py
s = str(input())
count = dict()
for x in s:
    count[x] = 1 + count.get(x,0)
tot = 0
n = len(s)
t = [0]*n
for x in count.keys():
    if count[x]%2 != 0:
        tot  = tot + 1
    if tot>1:
        break
if tot>1:
    print("NO SOLUTION")
else:
    i = 0
    for x in count.keys():
        if count[x] % 2 == 0:
            while( count[x]/2 > 0 ):
                t[i] = t[n-i-1] = x  
                count[x] = count[x] - 2
                i = i+1
        else:
            t[(n-1)//2] = x 
            if count[x]>1:
                for y in range(1,(count[x]+1)//2):
                    t[((n-1)//2) - y] = t[((n-1)//2) + y] = x
    print("".join(t))            
